File: hackerrank/problem solving/algorithms/2 - implementation/6 - subarray division/with morfo library/subarray_division.rb
# https://www.hackerrank.com/challenges/the-birthday-bar/problem?isFullScreen=true

require 'morfo'

def main
  Morfo.skip_input_ln
  chocolate_squares = Morfo.readln(&:to_i)
  day_month = Morfo.readln(&:to_i)
  puts birthday(chocolate_squares, day_month)
end

# n: length of chocolate_squares
# 1 <= n <= 100
# T: O(n) = O(100) = O(1)
# S: O(1) extra space
def birthday(chocolate_squares, day_month)
  day, month = day_month
  return 0 if month > chocolate_squares.size

  sum = chocolate_squares.lazy.take(month).sum
  count = sum == day ? 1 : 0

  count + (month...chocolate_squares.size).count do
    sum += chocolate_squares[_1] - chocolate_squares[_1 - month]
    sum == day
  end
end

main if __FILE__ == $0

File: hackerrank/problem solving/algorithms/2 - implementation/6 - subarray division/with morfo library/subarray_division.cs
// https://www.hackerrank.com/challenges/the-birthday-bar/problem?isFullScreen=true

using Morfo.IO;

Reader.SkipInputLn();
int[] chocolateSquares = Reader.ReadLn<int, int[]>();
int[] dayMonth = Reader.ReadLn<int, int[]>();
Console.WriteLine(Birthday(chocolateSquares, dayMonth));

// n: length of chocolateSquares
// 1 <= n <= 100
// T: O(n) = O(100) = O(1)
// S: O(1) extra space
int Birthday(int[] chocolateSquares, int[] dayMonth)
{
    int day = dayMonth[0];
    int month = dayMonth[1];

    if (month > chocolateSquares.Length)
        return 0;

    int sum = chocolateSquares.Take(month).Sum();
    int count = sum == day ? 1 : 0;

    for (int i = month; i < chocolateSquares.Length; i++)
    {
        sum += chocolateSquares[i] - chocolateSquares[i - month];
        if (sum == day)
            count++;
    }

    return count;
}
